### ISSUE TITLE
deadlock in MappingQEulerian

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#1829]
based on discussion on the mailing list, this looks like a deadlock in dealii::MappingQEulerian<3,
dealii::Vector<double>, 3>::compute_mapping_support_points:
```
#6 ScopedLock (m=..., this=) at
/home/oded/dealii/dealii_git_15102015/include/deal.II/base/thread_management.h:272
#7 dealii::MappingQEulerian<3, dealii::Vector,
3>::compute_mapping_support_points (this=0x99aa00, cell=...) at
/home/oded/dealii/dealii_git_15102015/source/fe/mapping_q_eulerian.cc:192
...
#33 0x00007ffff47f5a15 in dealii::MappingQEulerian<3, dealii::Vector,
3>::compute_mapping_support_points (this=0x99aa00, cell=...) at
/home/oded/dealii/dealii_git_15102015/source/fe/mapping_q_eulerian.cc:193
```

**Comment by tjhei** [comment]
I am attaching an example code 
[step-22.cc.txt](https://github.com/dealii/dealii/files/26618/step-22.cc.txt)
that hits that deadlock fairly consistently (dots stop appearing after a while).

@bangerth having looked at that code recently, do you understand why compute_mapping_support_points calls fe_values.reinit which in turn calls compute_mapping_support_points?

**Comment by bangerth** [comment]
I don't think I touched the logic of that class. My guess is that it works as it did when @luca-heltai or someone in his group wrote it.

I can look into it Monday when I'm back from Asia if nobody beats me to it.

**Comment by luca-heltai** [comment]
We never touched `MappinQEulerian`...

**Comment by kronbichler** [comment]
I had a short look at the relevant code. Here is a backtrace on the relevant thread:
```
#5  0x00007ffff2353db3 in dealii::Threads::Mutex::ScopedLock::ScopedLock (this=0x7fffd4221b20, m=...)
    at /home/kronbichler/deal/deal.II/include/deal.II/base/thread_management.h:272
#6  0x00007ffff302dc5e in dealii::MappingQEulerian<2, dealii::Vector<double>, 2>::MappingQEulerianGeneric::compute_mapping_support_points (this=0x9a6630, cell=...)
    at /home/kronbichler/deal/deal.II/source/fe/mapping_q_eulerian.cc:217
#7  0x00007ffff2fa452a in dealii::MappingQGeneric<2, 2>::transform_unit_to_real_cell (this=0x9a6630, cell=..., 
    p=...) at /home/kronbichler/deal/deal.II/source/fe/mapping_q_generic.cc:1223
#8  0x00007ffff2ff5325 in dealii::MappingQ<2, 2>::transform_unit_to_real_cell (this=0x8abc10, cell=..., p=...)
    at /home/kronbichler/deal/deal.II/source/fe/mapping_q.cc:473
#9  0x00007ffff078f007 in dealii::DataOut<2, dealii::DoFHandler<2, 2> >::build_one_patch (this=0x7fffffffc9d0, 
    cell_and_index=0x855970, data=..., patch=..., 
    curved_cell_region=dealii::DataOut<2, dealii::DoFHandler<2, 2> >::no_curved_cells, 
    patches=std::vector of length 1024, capacity 1024 = {...})
    at /home/kronbichler/deal/deal.II/source/numerics/data_out.cc:79
...
(lots of functions that call into DataOut::build_one_patch)
...
#23 0x00007ffff26a558e in tbb::task::wait_for_all (this=0x7fffd43f7a40) at /usr/include/tbb/task.h:744
#24 0x00007ffff26a8e9b in dealii::Threads::internal::TaskDescriptor<void>::join (this=0x7fffcc010b60)
    at /home/kronbichler/deal/deal.II/include/deal.II/base/thread_management.h:2888
#25 0x00007ffff26ab658 in dealii::Threads::Task<void>::join (this=0x7fffcc010b20)
    at /home/kronbichler/deal/deal.II/include/deal.II/base/thread_management.h:3001
#26 0x00007ffff26a8015 in dealii::Threads::TaskGroup<void>::join_all (this=0x7fffd4223b08)
    at /home/kronbichler/deal/deal.II/include/deal.II/base/thread_management.h:4009
#27 0x00007ffff2671fc1 in dealii::FESystem<2, 2>::compute_fill<2> (this=0x7fffffffcff8, mapping=..., cell=..., 
    face_no=4294967295, sub_no=4294967295, quadrature=..., mapping_internal=..., fedata=..., mapping_data=..., 
    output_data=...) at /home/kronbichler/deal/deal.II/source/fe/fe_system.cc:1263
#28 0x00007ffff2671ac5 in dealii::FESystem<2, 2>::fill_fe_values (this=0x7fffffffcff8, mapping=..., cell=..., 
    quadrature=..., mapping_internal=..., fe_data=..., mapping_data=..., output_data=..., 
    cell_similarity=dealii::CellSimilarity::none) at /home/kronbichler/deal/deal.II/source/fe/fe_system.cc:1024
#29 0x00007ffff290deb7 in dealii::FEValues<2, 2>::do_reinit (this=0x9a6850)
    at /home/kronbichler/deal/deal.II/source/fe/fe_values.cc:3763
#30 0x00007ffff2929f62 in dealii::FEValues<2, 2>::reinit<dealii::DoFHandler, false> (this=0x9a6850, cell=...)
    at /home/kronbichler/deal/deal.II/source/fe/fe_values.cc:3740
#31 0x00007ffff302dc81 in dealii::MappingQEulerian<2, dealii::Vector<double>, 2>::MappingQEulerianGeneric::compute_mapping_support_points (this=0x9a6630, cell=...)
    at /home/kronbichler/deal/deal.II/source/fe/mapping_q_eulerian.cc:218
#32 0x00007ffff2fa452a in dealii::MappingQGeneric<2, 2>::transform_unit_to_real_cell (this=0x9a6630, cell=..., 
    p=...) at /home/kronbichler/deal/deal.II/source/fe/mapping_q_generic.cc:1223
#33 0x00007ffff2ff5325 in dealii::MappingQ<2, 2>::transform_unit_to_real_cell (this=0x8abc10, cell=..., p=...)
    at /home/kronbichler/deal/deal.II/source/fe/mapping_q.cc:473
#34 0x00007ffff078f007 in dealii::DataOut<2, dealii::DoFHandler<2, 2> >::build_one_patch (this=0x7fffffffc9d0, 
    cell_and_index=0x854d58, data=..., patch=..., 
    curved_cell_region=dealii::DataOut<2, dealii::DoFHandler<2, 2> >::no_curved_cells, 
    patches=std::vector of length 1024, capacity 1024 = {...})
    at /home/kronbichler/deal/deal.II/source/numerics/data_out.cc:79
```
What happens is that DataOut::build_one_patch asks Mapping_Q_Eulerian for the support points. To this end, the underlying FEValues object is set up initialized, which internally goes to FESystem that spawns its own parallel tasks (frame 27). The TBB scheduler decides to go to another thread with that work; while waiting the current thread gets another slice of the outer for loop in DataOut and ends attempting to compute the mapping support points through Mapping_Q_Eulerian (frame 7). But it cannot get the lock since it is the very same thread to hold the lock. All other threads on the system also end up waiting for that same lock in some way or another.

As pointed out in #1830, spawning sub-tasks in `FESystem::compute_fill` is probably not a good idea anyway, which would fix this issue.

A more general conclusion seems to be that we can end up in a deadlock if code that is protected by a lock spawns its own sub-tasks that can get assigned somewhere else. Since we do not want to avoid nested parallelism altogether, the route forward must be to reduce the number of locks in cases where we have complex calling structures and only have locks for environments where we can make sure that no sub-parallelism is spawned.

**Comment by tjhei** [comment]
There is two interesting things here:
- The current thread is already holding the lock causing the deadlock. A different approach would be to use a critical section instead of a mutex.
- Why is compute_mapping_support_points calling compute_mapping_support_points and isn't that a bug too?

**Comment by bangerth** [comment]
On 11/04/2015 12:02 PM, Luca Heltai wrote:
> We never touched |MappinQEulerian|...

Ah, yes, my apologies. I confused this with `MappingFEField`. However, because 
they're very similar, what ails one is going to ail the other as well, I 
believe. Thinking about it, I believe I even pointed out this one problem 
during patch review :-)

**Comment by kronbichler** [comment]
> Why is compute_mapping_support_points calling compute_mapping_support_points and isn't that a bug too?

I assume you mean compute_mapping_support_points in frame 31 and compute_mapping_support_points in frame 6? This is due to TBB. While working on the outer method, TBB spawns sub-tasks. While waiting for the tasks to complete (likely all other threads are busy in similar code), TBB decides to put the current context to sleep (frame 23) and start another task context (it starts at frame 22 and continuous the parallel for loop for DataOut::build_patches). In that task, the same access to the lock appears again.

If anything, this would be a bug in TBB, but it really is a feature of the dynamic task scheduling. I see no way to avoid this unless we remove the locks in 'difficult' contexts.

> A different approach would be to use a critical section instead of a mutex.

I don't see how that would help because the underlying problem of only one processor calling fe_values.reinit() is still not addressed. The solution would be to replace the single FEValues object in MappingQEulerian to a ThreadLocalStorage<list<FEValues> > object similar to what we do with WorkStream, which would avoid the need for the lock. By the way, this would make the code more scalable with threads because right now the mapping support points are computed in serial (it's quite interesting that the data race appears so rarely given that the scheduler must be desperate to find work to do in parallel)...

**Comment by bangerth** [comment]
Correct, the only correct solution is to use a thread-local list. Fixing the issue with #1830 is really only going to paper over this.

**Comment by luca-heltai** [comment]
I think the only viable solution is what solved the problem for `MappingFEField`. I have not been able to sit down and do some real programming lately, but I'll try to figure if this is the same issue we had there. 

Moving the support points into the internal data (which is unique to each thread) is what solved it for `MappingFEField`. I have not looked at the details of `MappingQEulerian`, but I guess it should not be too different, as @bangerth suggested...

**Comment by kronbichler** [comment]
@luca-heltai I completely agree. This is the right solution and consistent with the interplay between mapping and the geometry that has been reworked by @bangerth .

**Comment by bangerth** [comment]
Yes, moving it into the internal data object is a much more elegant way to avoid the problem than using a threadlocal list of objects.

**Comment by bangerth** [comment]
@tjhei -- in your testcase, I imagine that one of the 10,000 invokations of `output_results` hangs? How long did you typically have to wait? It continues to run for me with the current setup, but I've removed some instances of parallelism recently.

**Comment by tjhei** [comment]
It only took a couple of seconds on my machine to hit the deadlock every time.

**Comment by tjhei** [comment]
I can no longer reproduce it, which is not surprising, because compute_fill looks different now.
